src/test/java/com/karankumar/bookproject/ui/book/BookFormTest.java

```diff
@@ -479,14 +517,169 @@ void shouldDisplaySeriesPosition_withSeriesPositionPopulated_whenBookHasSeriesPo
+        // given
+        bookForm = createBookForm(initialShelf, false);
+        bookForm.addListener(BookForm.SaveEvent.class, event -> bookService.save(event.getBook()));
+        bookForm.saveButton.click();
+        populateBookForm(newShelf, false, bookService.findAll().get(0));
+
+        // when
+        bookForm.saveButton.click();
+
+        // then
+        Assertions.assertEquals(1, bookService.count());
+        Book bookInDatabase = bookService.findAll().get(0);
+        assertBookInSpecificShelf(newShelf, bookInDatabase);
+    }
+
+    private void assertBookInSpecificShelf(ShelfName shelfName, Book bookInDatabase) {
+        // these values should be present no matter which shelf the book is in
+        Assertions.assertEquals(bookTitle, bookInDatabase.getTitle());
+        Assertions.assertEquals(shelfName, bookInDatabase.getPredefinedShelf().getPredefinedShelfName());
+        Assertions.assertEquals(firstName, bookInDatabase.getAuthor().getFirstName());
+        Assertions.assertEquals(lastName, bookInDatabase.getAuthor().getLastName());
+        Assertions.assertEquals(genre, bookInDatabase.getGenre());
+        Assertions.assertEquals(numberOfPages, bookInDatabase.getNumberOfPages());
+
+        // these values should only be present for specific predefined shelves
+        switch (shelfName) {
+            case TO_READ:
+                Assertions.assertNull(bookInDatabase.getDateStartedReading());
+                Assertions.assertNull(bookInDatabase.getDateFinishedReading());
+                Assertions.assertEquals(RatingScale.NO_RATING, bookInDatabase.getRating());
+                Assertions.assertNull(bookInDatabase.getPagesRead());
+                break;
+            case READING:
+                Assertions.assertEquals(dateStarted, bookInDatabase.getDateStartedReading());
+                Assertions.assertNull(bookInDatabase.getDateFinishedReading());
+                Assertions.assertEquals(RatingScale.NO_RATING, bookInDatabase.getRating());
+                Assertions.assertNull(bookInDatabase.getPagesRead());
+                break;
+            case READ:
+                Assertions.assertEquals(dateStarted, bookInDatabase.getDateStartedReading());
+                Assertions.assertEquals(dateFinished, bookInDatabase.getDateFinishedReading());
+                Assertions.assertEquals(0, bookInDatabase.getRating().compareTo(ratingVal));
+                Assertions.assertNull(bookInDatabase.getPagesRead());
+                break;
+            case DID_NOT_FINISH:
+                Assertions.assertEquals(dateStarted, bookInDatabase.getDateStartedReading());
+                Assertions.assertNull(bookInDatabase.getDateFinishedReading());
+                Assertions.assertEquals(0, bookInDatabase.getRating().compareTo(ratingVal));
+                Assertions.assertEquals(pagesRead, bookInDatabase.getPagesRead());
+                break;
```

Comment:
There is a bit of duplication here which can be removed if we want. Like we can create a private method that has these 4 asserts and based on the arguments (rating etc) passed we can assert the rating in that method and rest of the asserts are always same?

Reply:
I am not quite sure, if I understood you correctly. Do you mean something like this?
```java
private void assertStateSpecificFields(Book book, LocalDate dateStarted, LocalDate dateFinished,
                                           RatingScale rating, Integer pagesRead) {
        Assertions.assertEquals(dateStarted, book.getDateStartedReading());
        Assertions.assertEquals(dateFinished, book.getDateFinishedReading());
        Assertions.assertEquals(rating, book.getRating());
        Assertions.assertEquals(pagesRead, book.getPagesRead());
    }
```
And then calls like:
```java
assertStateSpecificFields(bookInDatabase, null, null, RatingScale.NO_RATING, null);
assertStateSpecificFields(bookInDatabase, dateStarted, dateFinished, ratingVal, null);
```

Reply:
yes I personally find this much better and just a minor comment but feel free to ignore. It would be more readable if we can create field variable for the nulls with proper names e.g. 
`assertStateSpecificFields(bookInDatabase, NULL_STARTED_DATE, NULL_FINISH_DATE, RatingScale.NO_RATING, NO_PAGES_READ)`
and `NULL_STARTED_DATE`, `NULL_FINISH_DATE`, `NO_PAGES_READ` can be final fields initialised to null

Reply:
I agree, the constants do increase the readability, so I added them.